core/templates/dev/head/exploration_editor/EditorServices.js

```diff
@@ -1839,3 +1890,293 @@ oppia.factory('explorationWarningsService', [
     };
   }
 ]);
+
+oppia.factory('lostChangesService', function() {
+  var CMD_ADD_STATE = 'add_state';
+  var CMD_RENAME_STATE = 'rename_state';
+  var CMD_DELETE_STATE = 'delete_state';
+  var CMD_EDIT_STATE_PROPERTY = 'edit_state_property';
+
+  var getRulesList = function(value) {
+    var rulesList = [];
+    value.rule_specs.forEach(function(ruleSpec) {
+      var ruleElm = angular.element('<div></div>');
+      ruleElm.append('<p>Type: ' + ruleSpec.rule_type + '</p>');
+      ruleElm.append(
+        '<p>Value: ' +
+          (
+            Object.keys(ruleSpec.inputs).map(function(input) {
+              return ruleSpec.inputs[input];
+            })
+          ).toString() +
+          '</p>');
+      rulesList.push(ruleElm);
+    });
+
+    return rulesList;
+  };
+
+  var getStatePropertyValue = function(
+    statePropertyValue) {
+    return angular.isArray(statePropertyValue) ?
+      statePropertyValue[statePropertyValue.length - 1] : statePropertyValue;
+  };
+
+  var isEmpty = function(obj) {
+    for (var property in obj) {
+      if (obj.hasOwnProperty(property)) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  var getRelativeChangeToGroups = function(changeObject) {
+    var newValue = changeObject.new_value;
+    var oldValue = changeObject.old_value;
+    var result = '';
+
+    if (angular.isArray(newValue) && angular.isArray(oldValue)) {
+      result = (newValue.length > oldValue.length) ?
+        'added' : (newValue.length === oldValue.length) ?
+        'edited' : 'deleted';
+    } else {
+      if (!isEmpty(oldValue)) {
+        if (!isEmpty(newValue)) {
+          result = 'edited';
+        } else {
+          result = 'deleted';
+        }
+      } else if (!isEmpty(newValue)) {
+        result = 'added';
+      }
+    }
+    return result;
+  };
+
+  var makeHumanReadable = function(lostChanges) {
+    var outerHtml = angular.element('<ul></ul>');
+    var statesWiseEditsMapping = {};
+
+    lostChanges.forEach(function(lostChange) {
+      var lostChangeEl = angular.element('<li></li>');
+      var lostChangeValue;
```

Comment:
The use of 'lostChangeEl' and 'lostChangeValue' is still a bit hard to follow, and in particular lostChangeValue seems to sometimes be a string and sometimes an element (line 2014). It should be possible to do away with both of these, and in each case just act directly to add a child element to the outerHtml. That will also help us maintain flexibility about the styles we want to use to display different types of change.